keps/sig-node/2837-pod-level-resource-spec/README.md

```diff
@@ -0,0 +1,1108 @@
+
+TBD. Do not review for the alpha stage.
+
+t
+### Test Plan
+
+[X] I/we understand the owners of the involved components may require updates to existing tests to make this code solid enough prior to committing the changes
+necessary to implement this enhancement.
+
+#### Unit tests
+
+This feature will touch multiple components. For alpha, unit tests coverage for following packages needs to be added:
+
+* Scheduler logic will be updated to consider pod-level requests. Hence pkg/scheduler will require additional coverage.
+* pkg/kubelet/cm will be updated to set pod-level cgroups using CPU requests and limits, and memory limits.
+* pkg/apis/core/validation/types_test.go and pkg/apis/core/validation since new fields are added in PodSpec and also new validation rules are required for the new fields.
+* pkg/kubeapiserver/admission for changes made in the admission logic for LimitRanger and ResourceQuota admission controllers.
+
+
+#### e2e tests
+
+Following scenarios need to be covered:
+
+
+
+* Cgroup settings when pod-level resources are set.
+* Validate scheduling and admission.
+* Validate the containers with no limits set are throttled on CPU when CPU usage reaches Pod level CPU limits.
+* Validate the containers with no limits set are OOMKilled when memory usage reaches Pod level memory limits.
+
+
+### Graduation Criteria
+
+
+#### Phase 1: Alpha (target 1.31)
+
+
+* Feature is disabled by default. It is an opt-in feature which can be enabled by setting the new fields in PodSpec.
+* Support the basic functionality for scheduler to consider pod-level resource requests to find a suitable node.
+* Support the basic functionality for kubelet to translate pod-level requests/limits to pod-level cgroup settings.
+* Unit test coverage.
+* E2E tests.
+* Documentation to announce the feature in alpha phase.
+
+
+#### Phase 2:  Beta (target 1.32)
+
+
+
+* User Feedback.
+* Feature is disabled by default. It is an opt-in feature which can be enabled by setting the new fields in PodSpec.
+* Support for all features that are scoped out of alpha phase in the KEP design.
+* Unit test coverage.
+* E2E tests.
+* Documentation update to announce feature in beta phase.
+* Benchmark tests for resource usage.
+
+
+#### GA (stable)
+
+
+* TBD
+
+
+### Upgrade / Downgrade Strategy
+
+The existing workloads will not have any impact since they won't be using the new field `resources` in PodSpec 
+at pod-level. If a cluster is upgraded, the cluster administrator will experience pre-upgrade behavior until
+`resources` field is specified at pod-level. 
+
+If the workloads are not using the new feature, downgrade won't be affected. However, if it is required to
+downgrade a cluster that has workloads using the pod-level limits functionality, following steps need to be
+performed:
+
+* Translate pod-level resource specifications to container-level specifications.
+
+* Delete existing workloads and recreate them with container-level specifications.
+
```

Comment:
if a user doesn't do these things, then they would loose their limits (as resources on pod spec would be stripped out as unknown by the API server). Is there anything to be done about that?

Reply:
@haircommander I think this is typical for any new feature/field. Do you have any suggestion?

Reply:
yeah I mean if we deemed this enough of an issue a fix would be https://github.com/kubernetes/enhancements/pull/4678#discussion_r1624946950 . I'm not feeling super stubborn about it but it is a concern.

Reply:
I think we can clearly document that toggling the feature may leave some pods in the undesired state, and to address the issues, users can delete existing pods. WDYT?